### ISSUE TITLE
run_translation_no_trainer with MBart: unsupported operand type(s) for /: 'dict' and 'int'

### ISSUE THREAD

**Issue by Doragd** [huggingface/transformers#13232]
## Environment info

- `transformers` version: 4.9.2
- Platform: Ubuntu 20.04
- Python version: 3.8
- PyTorch version (GPU?): 1.8.2
- Using GPU in script?: Yes
- Using distributed or parallel set-up in script?: Yes

In addition, I add the details of `accelerate` config.
```shell
In which compute environment are you running? ([0] This machine, [1] AWS (Amazon SageMaker)): 0
Which type of machine are you using? ([0] No distributed training, [1] multi-CPU, [2] multi-GPU, [3] TPU): 2
How many different machines will you use (use more than 1 for multi-node training)? [1]: 1
Do you want to use DeepSpeed? [yes/NO]: NO
How many processes in total will you use? [1]: 1
Do you wish to use FP16 (mixed precision)? [yes/NO]: yes
```

### Who can help

@patrickvonplaten, @patil-suraj

## Information

Model I am using (Bert, XLNet ...): facebook/mbart-large-cc25

The problem arises when using:
* [x] the official example scripts: examples/pytorch/translation/run_translation_no_trainer.py
* [ ] my own modified scripts: (give details below)

The tasks I am working on is:
* [ ] an official GLUE/SQUaD task: 
* [x] my own task or dataset: finetuning en-ro dataset of wmt16 translation

## To reproduce
I use the following script: note that i rename the `run_translation_no_trainer.py` with `run.py`
```shell
accelerate launch run.py \
    --model_name_or_path facebook/mbart-large-cc25 \
    --source_lang en \
    --target_lang ro \
    --dataset_name wmt16 \
    --dataset_config_name ro-en \
    --output_dir ~/tmp/tst-translation
```

## Expected behavior

![image](https://user-images.githubusercontent.com/26213546/130571764-92200a1f-222c-4267-a237-3b97e39c5e7c.png)

`outputs` is still an object of `Seq2SeqLMOutput`, however, the `outputs.loss` is a `dict` with keys `dict_keys(['loss', 'logits', 'past_key_values', 'encoder_last_hidden_state'])`

## Additional information
By the way, I also test another script： `examples/pytorch/text-classification/run_glue_no_trainer.py`
It met the same problem.
You can view it on colab: https://colab.research.google.com/drive/1BLt5rtHFdHaRliyqj_BKmM1_uzQclQHw?usp=sharing
![image](https://user-images.githubusercontent.com/26213546/130642391-f4f34bb0-e22c-4201-9382-0239b25a28e7.png)

**Comment by Doragd** [comment]
I have tested `outputs  = accelerator.unwrap_model(model)(**batch)`. It works well, and the `outputs.loss` is a `Tensor` as expected.

**Comment by Doragd** [comment]
I have checked that it is caused by `convert_to_fp32` in `accelerate/utils.py`.
```python
def convert_to_fp32(tensor):
    """
    Recursively converts the lements nested list/tuple/dictionary of tensors in FP16 precision to FP32.

    Args:
        tensor (nested list/tuple/dictionary of :obj:`torch.Tensor`):
            The data to convert from FP16 to FP32.

    Returns:
        The same data structure as :obj:`tensor` with all tensors that were in FP16 precision converted to FP32.
    """
    if isinstance(tensor, (list, tuple)):
        return honor_type(tensor, (convert_to_fp32(t) for t in tensor))
    elif isinstance(tensor, dict):
        return type(tensor)({k: convert_to_fp32(v) for k, v in tensor.items()})
    elif not hasattr(tensor, "dtype") or tensor.dtype != torch.float16:
        return tensor
    return tensor.float()
```

**Comment by Doragd** [comment]
To fix it，I modified `run_translation_no_trainer.py` with the following snippet in the training loop, and it works well.
```python
      outputs = model(**batch, return_dict=False)
      loss = outputs[0]
      loss = loss / args.gradient_accumulation_steps
      accelerator.backward(loss)
```

**Comment by Doragd** [comment]
By the way, I also test another script： `examples/pytorch/text-classification/run_glue_no_trainer.py`
It met the same problem.
You can view it on colab: https://colab.research.google.com/drive/1BLt5rtHFdHaRliyqj_BKmM1_uzQclQHw?usp=sharing
![image](https://user-images.githubusercontent.com/26213546/130642391-f4f34bb0-e22c-4201-9382-0239b25a28e7.png)

**Comment by Doragd** [comment]
For my solution, i think the `convert_to_fp32` should be corrected as follows:

```python
def convert_to_fp32(tensor):
    if isinstance(tensor, (list, tuple)):
        return honor_type(tensor, (convert_to_fp32(t) for t in tensor))
    elif isinstance(tensor, dict):
        return type(tensor)(**{k: convert_to_fp32(v) for k, v in tensor.items()}) # add **kwargs
    elif not hasattr(tensor, "dtype") or tensor.dtype != torch.float16:
        return tensor
    return tensor.float()
```
Do you think so? @patil-suraj

**Comment by Doragd** [comment]
@patrickvonplaten I have proposed a PR for this issue, could you help me check its correctness.
